statsmodels/regression/tests/test_recursive_ls.py

```diff
@@ -60,6 +63,175 @@ def test_endog():
+
+    # OLS residuals are equivalent to smoothed forecast errors
+    # (the latter are defined as e_t|T by Harvey, 1989, 5.4.5)
+    # (this follows since the smoothed state simply contains the
+    # full-information estimates of the regression coefficients)
+    actual = (mod.endog[:, 0] -
+              np.sum(mod['design', 0, :, :] * res.smoothed_state, axis=0))
+    assert_allclose(actual, res_glm.resid_response)
+
+    # Given the estimate of scale as `sum(v_t^2 / f_t) / (T - d)` (see
+    # Harvey, 1989, 4.2.5 on p. 183), then llf_recursive is equivalent to the
+    # full OLS loglikelihood (i.e. without the scale concentrated out).
+    desired = mod_glm.loglike(res_glm.params, scale=res_glm.scale)
+    assert_allclose(res.llf_recursive, desired)
+    # Alternatively, we can construct the concentrated OLS loglikelihood
+    # by computing the scale term with `nobs` in the denominator rather than
+    # `nobs - d`.
+    scale_alternative = np.sum((
+        res.standardized_forecasts_error[0, 1:] *
+        res.filter_results.obs_cov[0, 0]**0.5)**2) / mod.nobs
+    llf_alternative = np.log(norm.pdf(res.resid_recursive, loc=0,
+                                      scale=scale_alternative**0.5)).sum()
+    assert_allclose(llf_alternative, res_glm.llf)
+
+    # Prediction
+    # TODO: prediction in this case is not working.
+    if constraints is None:
+        design = np.ones((1, 3, 10))
+        actual = res.forecast(10, design=design)
+        assert_allclose(actual, res_glm.predict(np.ones((10, 3))))
+    else:
+        design = np.ones((2, 3, 10))
+        assert_raises(NotImplementedError, res.forecast, 10, design=design)
+
+    # Hypothesis tests
+    actual = res.t_test('m1 = 0')
+    desired = res_glm.t_test('m1 = 0')
+    assert_allclose(actual.statistic, desired.statistic)
+    assert_allclose(actual.pvalue, desired.pvalue, atol=1e-15)
+
+    actual = res.f_test('m1 = 0')
+    desired = res_glm.f_test('m1 = 0')
+    assert_allclose(actual.statistic, desired.statistic)
+    # TODO: Why is this tolerance relatively wide, whereas the t_test is very
+    # tight?
+    assert_allclose(actual.pvalue, desired.pvalue, atol=1e-6)
+
+    # Information criteria
+    # Note: the llf and llf_obs given in the results are based on the Kalman
+    # filter and so the ic given in results will not be identical to the
+    # OLS versions. Additionally, llf_recursive is comparable to the
+    # non-concentrated llf, and not the concentrated llf that is by default
+    # used in OLS. Compute new ic based on llf_alternative to compare.
+    # TODO: Why do I need a -1 here in the constrained case but not in the
+    # unconstrained case? It must have something to do with the fact that GLM
+    # does not report the intercept as a degree of freedom so we need to
+    # subtract it here, but then why don't we need to do that it the
+    # unconstrained case?
+    df_model = res.df_model - 1 if constraints is not None else res.df_model
```

Comment:
you can also directly check df_model for the example, i.e df_model == 1 in the constrained and ==2 in the unconstrained case

Reply:
Interesting. Maybe the df should be different in the RecursiveLS case with constraints; it's set here by the number of diffuse observations, which almost always turns out to be equal to the number of states with diffuse initialization, and so is same here regardless of the constraints.

Since the constraint (in this example) implies that there are is really only independent 1 state, I guess the best way to handle that would be with some kind of modification of the diffuse initialization to also enforce the constraint, but I don't know of a way to do that in the current structure.